### PR TITLE
Adding totp virtual devices p3

### PR DESCRIPTION
**Pull request by cviecco**

### CODE REVIEW

File: cmd/keymasterd/2fa_totp_test.go [hunk 1]
```diff
@@ -103,5 +105,225 @@ func TestGenerateNewTOTPSuccess(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(dir) // clean up
+	state.Config.Base.DataDirectory = dir
+	err = initDB(state)
+	if err != nil {
+		t.Fatal(err)
+	}
+	/// End of setup
```
Comment (rgooch) [inline cmd/keymasterd/2fa_totp_test.go]: s@///@//@

File: cmd/keymasterd/2fa_totp.go [hunk 2]
```diff
@@ -150,41 +154,12 @@ func (state *RuntimeState) GenerateNewTOTP(w http.ResponseWriter, r *http.Reques
...
-		logger.Printf("Wanted Post got='%s'", r.Method)
-		state.writeFailureResponse(w, r, http.StatusMethodNotAllowed, "")
-		return
-	}
-	err = r.ParseForm()
+	authUser, _, otpValue, err := state.commonTOTPPostHandler(w, r, state.getRequiredWebUIAuthLevel())
 	if err != nil {
-		logger.Println(err)
-		state.writeFailureResponse(w, r, http.StatusBadRequest, "Error parsing form")
+		logger.Printf("Error in common Handler")
```
[the remark this answers is not in the document]
Comment (cviecco) [reply cmd/keymasterd/2fa_totp.go]: No, the passing the eror to the client is not managed by the 'commonTOTPPostHandler'

File: cmd/keymasterd/2fa_totp.go [hunk 3]
```diff
@@ -253,3 +229,290 @@ func (state *RuntimeState) validateNewTOTP(w http.ResponseWriter, r *http.Reques
...
+		//http.Error(w, "db backend is offline for writes", http.StatusServiceUnavailable)
+		//return
+	}
+
+	//Check if value is on blacklist for that user?
+	// Check if there is a value successfully accepted for that counter value
+	const defaultPeriod = 30
+	counter := int64(math.Floor(float64(t.Unix()) / float64(defaultPeriod)))
+	if profile.LastSuccessfullTOTPCounter == counter {
+		logger.Printf("validateUserTOTP: alredy done TOTP for time period")
```
[the remark this answers is not in the document]
Comment (cviecco) [reply cmd/keymasterd/2fa_totp.go]: done

File: cmd/keymasterd/2fa_totp.go [hunk 4]
```diff
@@ -253,3 +229,290 @@ func (state *RuntimeState) validateNewTOTP(w http.ResponseWriter, r *http.Reques
...
+	case "text/html":
+		http.Redirect(w, r, profileURI(authUser, assumedUser), 302)
+	default:
+		w.WriteHeader(200)
+		fmt.Fprintf(w, "Success!")
+	}
+	return
+}
+
+func (state *RuntimeState) validateUserTOTP(username string, OTPValue int, t time.Time) (bool, error) {
```
[the remark this answers is not in the document]
Comment (cviecco) [reply cmd/keymasterd/2fa_totp.go]: added

File: cmd/keymasterd/2fa_totp.go [hunk 5]
```diff
@@ -253,3 +229,290 @@ func (state *RuntimeState) validateNewTOTP(w http.ResponseWriter, r *http.Reques
...
+
+///
+
+func (state *RuntimeState) commonTOTPPostHandler(w http.ResponseWriter, r *http.Request, requiredAuthLevel int) (string, int, int, error) {
+	// User must be logged in
+	if state.sendFailureToClientIfLocked(w, r) {
+		return "", 0, 0, errors.New("server still sealed")
+	}
+	/*
+	 */
```
[the remark this answers is not in the document]
Comment (cviecco) [reply cmd/keymasterd/2fa_totp.go]: done

File: cmd/keymasterd/2fa_totp.go [hunk 6]
```diff
@@ -253,3 +229,290 @@ func (state *RuntimeState) validateNewTOTP(w http.ResponseWriter, r *http.Reques
...
+		return "", 0, 0, errors.New("Invalid Method requeted")
+	}
+	err = r.ParseForm()
+	if err != nil {
+		logger.Println(err)
+		state.writeFailureResponse(w, r, http.StatusBadRequest, "Error parsing form")
+		return "", 0, 0, err
+	}
+	logger.Debugf(3, "Form: %+v", r.Form)
+
```
[the remark this answers is not in the document]
Comment (cviecco) [reply cmd/keymasterd/2fa_totp.go]: donw

**Pull request merged**